eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelFeatures.scala

```diff
@@ -135,4 +135,18 @@ object ChannelTypes {
     }
   }
 
+  /** If remote is requesting a non-default channel type, we accept it only if we support required features. */
+  def negotiateChannelType(localFeatures: Features, localChannelType: SupportedChannelType, remoteChannelType_opt: Option[ChannelType]): Either[ChannelType, SupportedChannelType] = remoteChannelType_opt match {
```

Comment:
Why it `remoteChannelType` optional? Shouldn't this function be used only when the remote provides a `channel_type`? And the switch to use `defautFromFeatures` otherwise should be done by the caller?

Reply:
Done with 2a6e652dc. I went with repeating the call to `ChannelTypes.defaultFromFeatures`, which makes it less confusing and looks closer to what we have in `Helpers.validateParamsFunder`.

Reply:
With the latest change, it becomes more apparent that what bothered me is rather that this isn't a negotiation at all: the function either returns `Left(remoteChannelType)` or `Right(remoteChannelType)`, just based on whether we're ok with their proposal or not.

It feels like this function should instead be:

```
def canUseChannelType(localFeatures: Features, remoteChannelType: ChannelType): Boolean
```

It becomes more similar to `canUseFeature` and makes more sense here, don't you think?

Reply:
It was going great, except that `negotiateChannelType` includes a cast from `ChannelType` to `SupportedChannelType` in the success case, which we lose with a `Boolean` return type. This doesn't compile:

```scala
val chosenChannelType: Either[InvalidChannelType, SupportedChannelType] = msg.channelType_opt match {
  // remote doesn't specify a channel type: we use spec-defined defaults
  case None => Right(ChannelTypes.defaultFromFeatures(d.localFeatures, d.remoteFeatures))
  // remote explicitly specifies a channel type: we negotiate
  case Some(remoteChannelType) if ChannelTypes.canUseChannelType(d.localFeatures, remoteChannelType) => Right(remoteChannelType)
  case Some(rejectedChannelType) => Left(InvalidChannelType(msg.temporaryChannelId, ChannelTypes.defaultFromFeatures(d.localFeatures, d.remoteFeatures), rejectedChannelType))
}

def canUseChannelType(localFeatures: Features, remoteChannelType: ChannelType): Boolean = remoteChannelType match {
    case _: UnsupportedChannelType => false
    // We ensure that we support the features necessary for this channel type.
    case proposedChannelType: SupportedChannelType => proposedChannelType.features.forall(f => localFeatures.hasFeature(f))
  }

```

Making the return type an `Option[SupportedChannelType]` isn't great either, because it doesn't convey the success/failure idea.

Reply:
Right, that's a good point...
It's not ideal, but I would prefer the solution where we return an `Option[SupportedChannelType]` to the `Either` case.
What about:

```scala
def supportChannelType(localFeatures: Features, remoteChannelType: ChannelType): Option[SupportedChannelType]
```

Which we would use like this:

```scala
case Some(remoteChannelType) => ChannelTypes.supportChannelType(d.localFeatures, remoteChannelType) match {
    case Some(...) => ...
    case None => ...
}
```

Reply:
Done with 355799ddc. I went with `areCompatible` instead of `supportChannelType`.